Add ERC: Nitrolite State Channels

mod opened the pull request:
### Pull Request Description

#### EIP Title
ERC-7824: State Channels Protocol Framework

#### Description
This ERC introduces a standard for implementing state channels on Ethereum. The proposal defines the essential structs and interfaces for state channels, enabling scalable off-chain execution, efficient on-chain dispute resolution with interoperability.

#### Author
Louis Bellet (<[mod@yellow.org](mailto:mod@yellow.org)>)

#### Status
Draft

#### Type
Standards Track (ERC)

#### Category
ERC

#### Build Status
All required builds have passed successfully. The proposal adheres to the provided guidelines and submission criteria.

#### Notes
This pull request creates a new draft EIP for ERC-7250 and follows the template provided in [[EIP-1](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1.md)](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1.md). The proposal aims to facilitate scalable state channel implementations on Ethereum while aligning with existing standards for account abstraction.

abcoathup commented on ERCS/erc-7250.md:
```diff
@@ -0,0 +1,160 @@
+---
+eip: 7250
```
```suggestion
eip: 7824
```
Assigning next sequential EIP/ERC/RIP number.
Numbers are assigned by editors & associates.

Please also update the filename.

mod replied:
I applied the change of number as suggested,
I decided to split into two ERC
- 7824: Define the EXIT format: https://github.com/statechannels/exit-format
- 7825: Define the State Channels types: https://docs.statechannels.org/protocol-tutorial/0010-states-channels/

Please note this code was audited by Consensys Diligence

dpatsora replied:
Hi!
We fixed the numbers

mod replied:
We can release number 7842

abcoathup commented on ERCS/erc-7250.md:
```diff
@@ -0,0 +1,160 @@
+---
+eip: 7250
+title: Nitro Protocol for State Channels with Account Abstraction
```
ERCs are for Ethereum standardization and not product/protocol promotion. 

```suggestion
title: State Channels with Account Abstraction
```
Can this ERC be used by any state channel implementation, or is it limited to Nitro based state channels?

mod replied:
Hello,
There is no commercial name or product nitro,
this can be used by most state-channels

We are from the open-source project https://statechannels.org/
https://github.com/statechannels

We are mentioned in https://ethereum.org/en/developers/docs/scaling/state-channels/

I will remove nitro, as it's an historical name of the open source protocol, but we also have the following name: "Stateful Asset Transfer Protocol"

dpatsora replied:
the title has been fixed

mod replied:
Suggested a new title

abcoathup commented on ERCS/erc-7250.md:
```diff
@@ -0,0 +1,160 @@
+title: Nitro Protocol for State Channels with Account Abstraction
+description: A standard for implementing interoperable state channels on Ethereum using the Nitro Protocol, integrated with ERC-4337 for enhanced account abstraction.
+author: Consensys Mesh (@statechennels), Layer 3 Foundation (@layer-3), Louis Bellet (@mod)
+discussions-to: https://github.com/ethereum/EIPs/discussions/7250
```
Please create a discussions topic in Eth Magicians with a link to this PR: https://ethereum-magicians.org/c/ercs/57

mod replied:
Account in progress, I will create the topic as soon as I have permissions

mod commented:
> The commit [7de0c81](https://github.com/ethereum/ERCs/commit/7de0c8193f0f698961d547b81c4bcdbb1c081094) (as a parent of [cfda43d](https://github.com/ethereum/ERCs/commit/cfda43d14346874d2ef31dd6fbed506a1716c75a)) contains errors. Please inspect the [Run Summary](https://github.com/ethereum/ercs/728/files) for details.

Will be corrected soon

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,146 @@
+---
+eip: 7824
+title: ERC for Exit Format in State Channels
```
```suggestion
title: Exit Format in State Channels
```

You don't need to mention ERC in the title.

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,146 @@
+
+## Abstract
+
+This ERC defines a standard **Exit Format** for specifying outcomes and asset allocations in state channels. The Exit Format supports diverse token types (native tokens, ERC-20, and others) and accommodates both simple and advanced allocation mechanisms. The standard ensures interoperability and clarity in the distribution of locked assets upon channel finalization, particularly in protocols like ...
```
I'd like to see a brief definition of what an "exit format" is in your abstract, for anyone who isn't familiar with the term.

nksazonov replied:
Fixed: e5b6887

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,146 @@
+
+This ERC introduces new structs and enums for use in state channel contracts. It is designed to be forward-compatible with existing protocols like Nitro. Implementations not using this standard will need to adapt their on-chain logic to parse Exit Format structures.
+
+## Test Cases
```
Test cases should be a list of inputs and expected outputs (or code essentially doing the same.)

You can omit this section, or you could include the input structs and what changes they should make on chain.

dpatsora replied:
fixed

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,146 @@
+
+## Rationale
+
+The Exit Format standardizes asset redistribution logic across state channel implementations, reducing ambiguity and improving interoperability. By supporting both simple and advanced allocations, it addresses a wide range of use cases, from basic payments to complex virtual channel setups.
```
This mostly reads like motivation because it's justifying the proposal as a whole. The rationale should be used to explain why you made certain technical choices within the document. For example, why you've included metadata fields or why `destination` is 32 bytes long. Obviously you should explain actually important stuff 🤣

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,146 @@
+
+State channels rely on clear and interoperable methods for redistributing locked assets when a channel closes. Current implementations lack a standardized format for defining these outcomes, leading to inefficiencies and integration challenges. This ERC introduces a unified **Exit Format** that can be adopted across state channel implementations, improving compatibility and simplifying on-chain i ...
+
+## Specification
```
Your specification doesn't really explain where these structs fit into the big picture, or explain why they need to be standardized separately from the interface that will be using them. Normally we only accept standards that define externally visible behavior (like events/functions.)

SamWilsn commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+---
+eip: 7825
```
```suggestion
eip: 7842
```

abcoathup replied:
Please also update the filename.

dpatsora replied:
FIxed the number and filename

mod replied:
Everything is now merged into 7824, 7842 is free to use.

SamWilsn commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+---
+eip: 7825
+title: State Channel Standard data types and interfaces
```
```suggestion
title: State Channel Data Types & Interfaces
```

dpatsora replied:
Fixed

SamWilsn commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+---
+eip: 7825
+title: State Channel Standard data types and interfaces
+description: Interoperable state channels on Ethereum using State Channel Protocol, integrated with ERC-4337 for enhanced account abstraction.
```
```suggestion
description: Interoperable state channels using State Channel Protocol, integrated with ERC-4337 for enhanced account abstraction.
```

dpatsora replied:
Fixed

mod replied:
Updated once more

SamWilsn commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+
+### Integration with ERC-4337
+
+#### Benefits
```
Anything not describing requirements should be in the rationale or motivation section.

dpatsora replied:
fixed

SamWilsn commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+
+## Rationale
+
+This ERC builds on the modularity of Nitro Protocol and the flexibility of ERC-4337, creating a unified standard for state channel implementations. By decoupling execution rules from core protocol logic, it ensures adaptability for various applications while maintaining a secure and efficient dispute resolution process.
```
This reads more like motivation. See my earlier comment on motivation vs. rationale.

SamWilsn commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+
+This ERC depends on ERC-4337 and assumes its availability for account abstraction functionality. Legacy state channel implementations without Nitro compatibility will require adaptation.
+
+## Test Cases
```
Same about test cases

dpatsora replied:
Fixed

SamWilsn reviewed (changes requested):
Need to change the second EIP number.

abcoathup commented on ERCS/erc-7825.md:
```diff
@@ -0,0 +1,160 @@
+title: State Channel Standard data types and interfaces
+description: Interoperable state channels on Ethereum using State Channel Protocol, integrated with ERC-4337 for enhanced account abstraction.
+author: State Channels (@statechannels), Layer-3 Foundation (@layer-3), Louis Bellet (@mod)
+discussions-to: https://ethereum-magicians.org/c/ercs/57
```
Please create a discussions topic in Eth Magicians linking to this PR
https://ethereum-magicians.org/c/ercs/57

mod replied:
done
https://ethereum-magicians.org/t/erc-7824-state-channels-framework/22566

mod commented:
We have decided to merge the ERC's into one and I will post the request for comments to Ethereum Magician once done.

mod committed:
Adding proposed ERC-7824

mod committed:
Update discussion link

mod commented:
Changes:
- Merged both ERCs into ERC-7824 `State Channel Framework`
- Added all Interfaces and data types
- Added an example
- Rewrote the Rational
- Created and linked the discussion: https://ethereum-magicians.org/t/erc-7824-state-channels-framework/22566

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,354 @@
+4. **Dispute Resolution**: If a disagreement arises, a participant can force a state on-chain.
+5. **Finalization**: Upon agreement or after a timeout, the channel is finalized, and the outcome is settled.
+
+#### Example Application
```
I'd put the example application into the Reference Implementation section.

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,354 @@
+
+
+
+## Rationale
```
None of these points belong in Rationale. This section is for explaining technical choices made within the proposal itself, but your content is justifying the proposal as a whole. My favorite analogy is:

> Motivation: we need to build a shed because...
> Rationale: we painted the shed red because...

SamWilsn commented on ERCS/erc-7824.md:
```diff
@@ -0,0 +1,354 @@
+This ERC is agnostic of the protocol rules that must be implemented using IForceMoveApp. While the smart-contract framework is a simple set of convention, much of the security risk is moved off-chain. Protocols using State channels must perform security audit of their client and server backend implementation.
+
+## Copyright
+Copyright and related rights waived via [CC0](https://github.com/ethereum/ERCs/blob/master/LICENSE.md).
```
Please use the exact waiver from the template.

mod commented:
@SamWilsn thanks for coming back at us, we are working on this SDK and will provide an updated version middle of next week with resolved comments.

mod committed:
Updating nitro to nitrolite version

Commit:
Update ERCS/erc-7824.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

Commit:
Update ERCS/erc-7824.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

Commit:
Update ERCS/erc-7824.md

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

Commit:
Apply suggestions from code review

Co-authored-by: Sam Wilson <57262657+SamWilsn@users.noreply.github.com>

Commit:
Merge branch 'master' into master

SamWilsn reviewed (commented):
Looking good! Just a few changes left from my previous review.
